kubeflow/trainer/rhai/traininghub.py

```
@@ -139,23 +144,427 @@ def _compose_exec_script(func_code: str, func_file: str) -> str:
+        """Read last line of JSONL file (most recent metrics from rank 0)."""
+        # Determine file pattern based on algorithm
+        if self.algorithm == "osft":
+            # Mini-Trainer: Read from rank 0 file explicitly
+            metrics_file = f"{{self.ckpt_output_dir}}/{osft_metrics_file}"
+        else:  # sft
+            # InstructLab: Find rank 0 file (global0.jsonl)
+            pattern = f"{{self.ckpt_output_dir}}/{sft_metrics_pattern}"
+            files = glob.glob(pattern)
+
+            if not files:
+                return {{}}
+
+            # Prefer rank 0 file
+            rank_0_files = [f for f in files if '{sft_metrics_file}' in f]
+            metrics_file = rank_0_files[0] if rank_0_files else files[0]
+
+        # Read configuration from first line (SFT) or training_params.json (OSFT)
+        # and metrics from last line
+        try:
+            if not os.path.exists(metrics_file):
+                return {{}}
+
+            config = {{}}
+            last_line = None
+
+            # For OSFT (Mini-Trainer), read config from training_params.json
+            if self.algorithm == "osft":
+                config_file = f"{{self.ckpt_output_dir}}/{osft_config_file}"
+                if os.path.exists(config_file):
+                    try:
+                        with open(config_file, 'r') as f:
+                            config = json.load(f)
+                    except:
+                        pass
+
+            # Read metrics from JSONL
+            with open(metrics_file, 'r') as f:
+                for i, line in enumerate(f):
+                    if line.strip():
+                        # For SFT (InstructLab), first line contains training configuration
+                        if self.algorithm == "sft" and i == 0:
+                            try:
+                                config = json.loads(line)
+                            except:
+                                pass
```

We should log errors here too rather than letting it fail silently